bootstrap-webpack/bootstrap/spasm/modules/spa.js

```diff
@@ -66,6 +66,7 @@ export let jsExports = {
         innerText: (nodePtr,textLen, textOffset) => {
             nodes[nodePtr].innerText = decoder.string(textLen,textOffset);
         },
+        setAttributeBool: setAttribute,
```

Comment:
is this right actually? shouldn't a false value remove the attribute and a true value set it to `attributeName="attributeName"`?

Reply:
Nah, that behaviour is for properties. Like 'autofocus', 'required' etc. You should not use @attr for those.

Reply:
Spasm was written before I wrote the webidl bindings. With them I could merge the @prop and @attr into one, and use introspection on the node type to determine which are properties and attributes.

Reply:
but wouldn't attr also be used as a kind of DOM API when reusing components?

Reply:
I don't quite follow.

Reply:
well assuming you could use a D webasm component in isolation in HTML and JavaScript, like making some `<LineChart>` component in D: I would expect once supporting that, adding `@attr` fields would be a way of input for values, like in custom components in JS right now.

Also for interop with JS frameworks which use attributes this would make sense

Reply:
You mean with web-components/custom-elements? I believe they have callbacks when props/attrs change so you can call some arbitrary D wasm code to update the values (with the benefit that any value dependencies you have get updated as well).

But then the D wasm code wouldn't control the outer element, just the inner ones.

So you can have react create a `<linechart width={x} height={x}/>` element in which D code would render a div, canvas, etc. inside of it. The width and height attributes would be written by react, and read by D via custom-element's `attributeChangedCallback`.

Reply:
wouldn't you use `@attr` for this purpose then? When the component updates its own width/height and outside in JS you are later reading that attribute from your custom component.

I think this is a good reason to keep `@attr`

Reply:
Nah, you would use a callback to get information back from the component to the outside. Like 

```js
const LineChart = () => {
  return (<linechart width={500} height={400} onChange={(data)=>{
    // do stuff with data
  }}/>);
}
```

I don't want to get rid of @attr, I want to merge the 2 UDA's together and use introspection to determine if something is an attribute or a property. Mostly because understanding the difference is very hard. For instance, is `innerHTML` an @attr or a @prop? What about `checked`, `value`, `cellpadding`? They are worse than irregular verbs; the distinction looks historical and arbitrary.

Reply:
I think what's more confusing is that `@prop` can affect the DOM, like checked has a getter and setter that read/write to the DOM. All `@attr` affects the DOM and is consistent in that but not all `@prop` may affect the DOM.

Reply:
It is a mess.

`checked` is officially an attribute.

To create one you can write:
`<input type="checkbox" checked>` or
`<input type="checkbox" checked="anything-goes">`

To create one programatically you need to use `setAttribute('checked', true);` and `removeAttribute('checked');`, calling `setAttribute('checked', false)` still makes it checked.

Another option is to use `toggleAttribute('checked')` which will do what you expect.

But there is also a checked property, where you can simply do `input.checked = true` and `input.checked = false`.

Setting the property will not change the HTML, but the input element will visually reflect it regardless. On top of that the property takes precedence. So setting `input.checked = true || false` will overwrite whatever `setAttribute('checked', true)` or `removeAttribute('checked')` you might have done, resulting in the possibility of having a `<input checked="true">` that visually isn't checked.

For that reason I decided that checked should be set like a property (e.g. `input.checked = true`). Also, I want `@attr myAttrs = false` to mean `myAttrs="false"` instead of calling `removeAttribute("myAttrs")`, which I would need to do if I did checked as an attr.

To top it off webidl calls them attributes https://github.com/skoppe/spasm/blob/master/webidl/definitions/html.idl#L690, and has no notion of properties at all. At the time I checked if there was a standard definition of attributes and properties. I thought I came close with "attributes have values and properties don't", but alas, that is only true in some domains.

Same with autofocus, although the docs are clearer on that one:

> The HTMLSelectElement.autofocus property is a Boolean that reflects the autofocus HTML attribute, which indicates whether the associated <select> element  will get input focus when the page loads, unless the user overrides it.

Also, since react uses the props terminology, I figured attributes have lost.